File: solutions/647/main.py
from typing import List 

class Solution:
    def countSubstrings(self, s: str) -> int:
        result = 0
        # dp table
        # i, j means s[i:j] is palindrome
        dp = [[False] * (len(s)) for _ in range(len(s))]
        # initialize
        for i in range(len(s)):
            # case "a"
            dp[i][i] = True
            result += 1
        # traverse dp table
        for i in range(len(s)-1, -1, -1):
            for j in range(i+1, len(s)):
                if s[i] != s[j]:
                    # this cannot be palindrome
                    pass
                else:
                    if j == i+1:
                        # case "aa"
                        dp[i][j] = True
                        result += 1
                    else:
                        dp[i][j] = dp[i+1][j-1]
                        if dp[i][j]:
                            result += 1
        return result

if __name__ == "__main__": 
    s = "aaa"
    sol = Solution()
    assert sol.countSubstrings(s) == 6

File: solutions/647/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sat Nov 27 11:43:35 2021
> Description:   
 ************************************************************************/
#include <numeric> 
#include <vector> 
#include <string> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int countSubstrings(string s) {
        // dp table
        vector<vector<int>> dp(s.size(), vector<int>(s.size(), 0));
        // initialize
        for (int i = 0; i < s.size(); ++i) dp[i][i] = 1;
        // traverse dp table
        for (int i = s.size()-1; i >= 0; --i) {
            for (int j = i+1; j < s.size(); ++j) {
                if (s[i] == s[j]) {
                    if (i == j-1) dp[i][j] = 1;
                    else dp[i][j] = dp[i+1][j-1];
                }
            }
        }
        // get result
        int sum = 0;
        for (auto& row : dp) {
            sum = accumulate(row.begin(), row.end(), sum);
        }
        return sum;
    }
};

TEST(Test647, SimpleTest) {
    string s = "aaa"; 
    Solution sol; 
    EXPECT_EQ(sol.countSubstrings(s), 6);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
